apm-agent-plugins/apm-javalin-plugin/src/main/java/co/elastic/apm/agent/javalin/JavalinInstrumentation.java

```diff
@@ -0,0 +1,167 @@
+    public String getAdviceClassName() {
+        return "co.elastic.apm.agent.javalin.JavalinInstrumentation$HandlerAdapterAdvice";
+    }
+
+    public static class HandlerAdapterAdvice {
+
+        private static Field handlerTypeField = null;
+
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static Object setSpanAndTransactionName(@Advice.This Handler handler,
+                                                       @Advice.Argument(0) Object ctxObject) throws Exception {
+            final Transaction transaction = tracer.currentTransaction();
+            if (transaction == null) {
+                return null;
+            }
+            Context ctx = ((Context) ctxObject);
+
+            // TODO FIXME I don't think this is the right way, I am sure there is a fancy bytecode trick that I am not aware off
+            if (handlerTypeField == null) {
+                Field field = ctx.getClass().getDeclaredField("handlerType");
+                field.setAccessible(true);
+                handlerTypeField = field;
+            }
+
+            final HandlerType handlerType = (HandlerType) handlerTypeField.get(ctx);
+            final String handlerClassName = handler.getClass().getName();
+
+            // transaction name like the HTTP path
+            if (handlerType.isHttpMethod()) {
+                transaction.setFrameworkName(FRAMEWORK_NAME);
+                transaction.setFrameworkVersion(VersionUtils.getVersion(Handler.class, "io.javalin", "javalin"));
+                // no need for anonymous handler class names in the transaction
+                if (handlerClassName.startsWith("co.elastic.apm.agent.javalin.JavalinHandlerLambdaInstrumentation$WrappingHandler")
+                    || handlerClassName.startsWith("io.javalin.http.JavalinServlet")) {
+                    setName(transaction, handlerType.name() + " " + ctx.endpointHandlerPath());
+                } else {
+                    setName(transaction, handlerType.name() + " " + ctx.endpointHandlerPath() + " " + handlerClassName);
+                }
+            }
+
+            // do not create an own span for JavalinServlet.addHandler, as this not added by the users code and leads to confusion
+            if (handlerClassName.startsWith("io.javalin.http.JavalinServlet")) {
+                return null;
+            }
+
+            // create own span for all handlers including after/before
+            final Span span = startSpan(handlerType.name(), ctx.matchedPath(), ctx.method(), ctx.url());
+            return span;
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+        public static void onAfterExecute(@Advice.Enter @Nullable Object spanObj,
+                                          @Advice.Argument(0) Object ctxObject,
+                                          @Advice.Thrown @Nullable Throwable t) {
+            if (spanObj instanceof Span) {
```

Comment:
I don't think this check is necessary. It should be enough to check `if (spanObj != null)`